python/lsst/daf/persistence/butler.py

```diff
@@ -136,44 +175,213 @@ def __init__(self, root, mapper=None, **mapperArgs):
+                                         only be read from used if the tag in the dataId matches a tag used
+                                         for that repository.
+        """
+        if butlerIOParents is None:
+            butlerIOParents = {}
+
+        tags = copy.copy(setify(tags))
+        tags.update(args.tags)
+
+        # If this repository has already been loaded/created, compare the input args to the exsisting repo;
+        # if they don't match raise an exception - the caller has to sort this out.
+        if args.cfgRoot in self._repos.byCfgRoot:
+            repoData = self._repos.byCfgRoot[args.cfgRoot]
+            if not repoData.cfg.matchesArgs(args):
+                raise RuntimeError("Mismatched repository configurations passed in or loaded:" +
+                                   "\n\tnew args:%s" + 
+                                   "\n\texisting repoData:%s" %
+                                   (args, repoData))
+            repoData.tags.update(setify(tags))
+        else:
+            # If we are here, the repository is not yet loaded by butler. Do the loading:
+
+            # Verify mode is legal (must be writeable for outputs, readable for inputs).
+            # And set according to the default value if needed.
+            if inout == 'out':
+                if 'w' not in args.mode:
+                    raise RuntimeError('Output repositories must be writable.')
+            elif inout == 'in':
+                if 'r' not in args.mode:
+                    raise RuntimeError('Input repositories must be readable.')
+            else:
+                raise RuntimeError('Unrecognized value for inout:' % inout)
+            
+            # Try to get the cfg.
+            # If it exists, verify no mismatch with args.
+            # If it does not exist, make the cfg and save the cfg at cfgRoot.
+            cfg = Storage.getRepositoryCfg(args.cfgRoot)
+
+            parentsToAdd = []
+            if cfg is not None:
+                if inout == 'out':
+                    # Parents used by this butler instance must match the parents of any existing output 
+                    # repositories used by this butler. (If there is a configuration change a new output 
+                    # repository should be created). 
+                    # IE cfg.parents can have parents that are not passed in as butler inputs or readable 
+                    # outputs, but the butler may not have readable i/o that is not a parent of an already 
+                    # existing output.
+                    for cfgRoot in butlerIOParents:
+                        if cfgRoot not in cfg.parents:
+                            if cfgRoot != args.cfgRoot or 'r' not in args.mode:
+                                raise RuntimeError(
+                                    "Existing output repository parents do not match butler's inputs.")
```

Comment:
I don't understand `or 'r' not in args.mode` here.  If we're setting up an output repository, shouldn't this check do the same thing regardless of whether or not it's readable as well as writeable?